File: resources/aci_api/Java/queueInfo_getStatus/src/main/java/com/autonomy/aci/samples/QueueinfoGetstatus.java
package com.autonomy.aci.samples;

import com.autonomy.aci.client.services.AciServiceException;
import com.autonomy.aci.client.services.AciErrorException;
import com.autonomy.aci.client.services.AciService;
import com.autonomy.aci.client.services.impl.AciServiceImpl;
import com.autonomy.aci.client.services.impl.DocumentProcessor;
import com.autonomy.aci.client.transport.AciParameter;
import com.autonomy.aci.client.transport.AciServerDetails;
import com.autonomy.aci.client.transport.InputStreamActionParameter;
import com.autonomy.aci.client.transport.impl.AciHttpClientImpl;
import com.autonomy.aci.client.transport.impl.BteaEncryptionCodec;
import com.autonomy.aci.client.util.ActionParameters;
import org.apache.http.impl.client.HttpClientBuilder;
import org.w3c.dom.Document;
import java.io.StringWriter;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathFactory;
import org.xml.sax.SAXException;
import java.util.logging.Logger;

import java.util.Optional;

public class QueueinfoGetstatus {

	private static final Logger LOG = Logger.getLogger(QueueinfoGetstatus.class.getName());
	private static boolean USE_IDOL_OEM_ENCRYPTION = true;
	private static final String IDOL_OEM_ENCRYPTION_KEY_ENV_VAR = "ACI_API_TUTORIAL_PROGRAM_ENCRYPTION_KEY";
	private static final String IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE = "REPLACE_WITH_MY_LONG_KEY_STRING";
	private static final String IDOL_OEM_ENCRYPTION_KEY = IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE;

	static class Args {
        static final int EXPECTED_MIN_NUM_ARGS = 3;
        static final int EXPECTED_MAX_NUM_ARGS = 4;
		final String aciHost;
		final Integer aciPort;
		final String queueName;
		Optional<String> token = Optional.empty();
		
		//Prints program usage information to standard output
		private void displayusageinfo()
		{
			System.console().printf("Usage: QueueinfoGetstatus <HOSTNAME/IP> <ACI_PORT_#> <QUEUE_NAME> [<TOKEN>]\n");
			System.console().printf("Parameters:\n");
			System.console().printf("    <HOSTNAME/IP>	Hostname / IP address of ACI service.\n");
			System.console().printf("    <ACI_PORT_#>	ACI port number of ACI service.\n");
			System.console().printf("    <QUEUE_NAME>	Queue name - e.g. fetch.\n");
			System.console().printf("    [<TOKEN>]		Optional. Asynchronous action token.\n");
		}


		Args(final String[] stringArgs) throws IllegalArgumentException {
			if ((stringArgs.length < EXPECTED_MIN_NUM_ARGS) && (stringArgs.length > EXPECTED_MAX_NUM_ARGS)) {
				displayusageinfo();
				throw new IllegalArgumentException("Wrong number of arguments (three required, fourth is optional).");
			}
			
			aciHost = stringArgs[0];
			aciPort = Integer.parseInt(stringArgs[1]);
			queueName = stringArgs[2];

			System.console().printf("ACI Host: %s\n", aciHost);
			System.console().printf("ACI Port: %d\n", aciPort);
			System.console().printf("Queue Name: %s\n", queueName);
			if (stringArgs.length == EXPECTED_MAX_NUM_ARGS)
			{
				token = Optional.of(stringArgs[3]);
				System.console().printf("Token: %s\n", token);
			}
		}
	}

	private static String getStringFromDocument(Document doc) throws Exception
	{
		StringWriter writer = new StringWriter();
		StreamResult result = new StreamResult(writer);
		
		TransformerFactory tf = TransformerFactory.newInstance();
		Transformer transformer = tf.newTransformer();
		transformer.setOutputProperty(OutputKeys.INDENT, "yes");
		transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
		transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");		
		transformer.transform(new DOMSource(doc), result);

		return writer.toString();
	}

    private void queueinfo_getstatus(String[] args) throws Exception
	{
		Args parsedArgs;
		final AciService aciService;
		final ActionParameters parameters;
		final Document response;

		try {
			parsedArgs = new Args(args);
		} catch(Exception e) {
			LOG.severe("Unable to parse arguments.\n");
			throw e;
		}
		
		try {
			final AciServerDetails aciServerDetails = new AciServerDetails(parsedArgs.aciHost, parsedArgs.aciPort);
			final BteaEncryptionCodec encryptionCodec;

			// Initialize OEM licensing
			if (USE_IDOL_OEM_ENCRYPTION)
			{
				String key = System.getenv(IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);

				if (key != null)
				{
					LOG.info("Using OEM encryption key from environment variable: " + IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);
					if (key.equals(IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE)) {
						LOG.info("Don't forget to correctly set " + IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);
						throw new Exception("Invalid OEM encryption key");
					} else {
						encryptionCodec = new BteaEncryptionCodec(key);
					}
				} else {
					LOG.info("Using OEM encryption key from constant IDOL_OEM_ENCRYPTION_KEY");
					if (IDOL_OEM_ENCRYPTION_KEY.equals(IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE)) {
						LOG.info("Don't forget to correctly set constant: IDOL_OEM_ENCRYPTION_KEY");
						throw new Exception("Invalid OEM encryption key");
					} else {
						encryptionCodec = new BteaEncryptionCodec(IDOL_OEM_ENCRYPTION_KEY);
					}
				}

				aciServerDetails.setEncryptionCodec(encryptionCodec);				
			}
		
			aciService = new AciServiceImpl(new AciHttpClientImpl(HttpClientBuilder.create().build()), aciServerDetails);
		} catch(Exception e) {
			LOG.severe("Unable to set up AciService.\n");
			throw e;
		}

		try {
			// Set action parameters
			parameters = new ActionParameters("QUEUEINFO");
			parameters.add("queueAction", "getStatus");
			parameters.add("queueName", parsedArgs.queueName);
			if (parsedArgs.token.isPresent())
			{
				parameters.add("token", parsedArgs.token.get());
			}
			parameters.add("responseFormat", "xml");
		} catch(Exception e) {
			LOG.severe("Unable to set up ActionParameters.\n");
			throw e;
		}

		try {
			LOG.info("Executing action...");

			response = aciService.executeAction(parameters, new DocumentProcessor());
		} catch(Exception e) {
			LOG.severe("Unable to execute action.\n");
			throw e;
		}

		try {
			LOG.info("Handling response...");

			String responseStr = getStringFromDocument(response);
			System.console().printf("\n" + responseStr + "\n");
			
			// Get <response> and print it
			final XPath xpath = XPathFactory.newInstance().newXPath();

			System.console().printf("response: %s\n", xpath.evaluate("/autnresponse/response", response));

			if (parsedArgs.token.isPresent())
			{
				System.console().printf("Queue Name: %s.\nToken: %s.\nStatus: %s\n", parsedArgs.queueName, parsedArgs.token.get(), xpath.evaluate("/autnresponse/responsedata/actions/action/status", response));
			}
		} catch(Exception e) {
			LOG.severe("Unable to handle response.\n");
			throw e;
		}
    }

    public static void main(String[] args) throws Exception
    {
		LOG.info("Program loaded");

		QueueinfoGetstatus obj = new QueueinfoGetstatus();

        try {
            obj.queueinfo_getstatus(args);
        } catch(Exception e) {
            LOG.severe("Scary exception happened: " + e.getMessage());
            throw e;
        }
    }	
}

File: resources/aci_api/C/src/queueInfo_getstatus/queueInfo_getStatus.c
#include "client.h"
#include "client_defs.h"
#include "queueInfo_getStatus.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>


//Prints program usage information to standard output
void displayusageinfo()
{
	printf("Usage: getStatus <HOSTNAME/IP> <ACI_PORT_#> <QUEUE_NAME> [<TOKEN>]\n");
	printf("Parameters:\n");
	printf("    <HOSTNAME/IP>	Hostname / IP address of ACI service.\n");
	printf("    <ACI_PORT_#>	ACI port number of ACI service.\n");
	printf("    <QUEUE_NAME>	Queue name - e.g. fetch.\n");
	printf("    [<TOKEN>]		Optional. Asynchronous action token.\n");
}

int main(int argc, char** argv)
{
	printf("Program loaded.\n");
	
	if (argc == EXPECTED_MIN_NUM_ARGS)
	{
		printf("ACI Host: %s\n", argv[1]);
		printf("ACI Port: %s\n", argv[2]);
		printf("Queue Name: %s\n", argv[3]);
	}
	else if (argc == EXPECTED_MAX_NUM_ARGS)
	{
		printf("ACI Host: %s\n", argv[1]);
		printf("ACI Port: %s\n", argv[2]);
		printf("Queue Name: %s\n", argv[3]);
		printf("Token: %s\n", argv[4]);
	}
	else
	{
		printf("Error: Program requires at least three arguments.\n");
		displayusageinfo();
		exit(1);
	}

	aciInit();

	// Create the connection object
	t_aciObject* pConnection = NULL;
	aciObjectCreate(&pConnection, ACI_CONNECTION);

	// Set host details
	aciObjectParamSetString(pConnection, ACI_HOSTNAME, argv[1]);
	aciObjectParamSetInt(pConnection, ACI_PORTNUMBER, atoi(argv[2]));
	aciObjectParamSetInt(pConnection, ACI_CONN_RETRIES, 3);

	aciObjectToString(pConnection);
	printf("Queue Name: %s\n", argv[3]);

	if (argc == EXPECTED_MAX_NUM_ARGS)
	{
		printf("Token: %s\n", argv[4]);
	}
	
	// Initialize OEM licensing
	if (USE_IDOL_OEM_ENCRYPTION)
	{
		/* TODO: Look at using Windows only _dupenv_s() and then turn-off _CRT_SECURE_NO_WARNINGS */
		char *key = getenv(IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);

		if (key != NULL)
		{
			printf("\nUsing OEM encryption key from environment variable: %s\n", IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);
			if (!strcmp(key, IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE))
			{
				printf("Don't forget to correctly set %s\n", IDOL_OEM_ENCRYPTION_KEY_ENV_VAR);
				printf("Invalid OEM encryption key\n");
				exit(1);
			} else {
				aciInitEncryption(TRUE, "TEA", key);
			}
		} else {
			printf("\nUsing OEM encryption key from getStatus.h\n");
			if (!strcmp(IDOL_OEM_ENCRYPTION_KEY, IDOL_OEM_ENCRYPTION_KEY_STUB_VALUE)) {
				printf("Don't forget to correctly set constant: IDOL_OEM_ENCRYPTION_KEY\n");
				printf("Invalid OEM encryption key\n");
				exit(1);
			} else {
				aciInitEncryption(TRUE, "TEA", IDOL_OEM_ENCRYPTION_KEY);
			}
		}
	}

	// Create the command object
	t_aciObject* pCommand = NULL;
	aciObjectCreate(&pCommand, ACI_COMMAND);

	// Set command to execute
	aciObjectParamSetString(pCommand, ACI_COM_COMMAND, "QUEUEINFO");
	
	// Set http method
	aciObjectParamSetBool(pCommand, ACI_COM_USE_POST,FALSE);

	// Set action parameters
	aciObjectParamSetString(pCommand, "queueName", argv[3]);
	aciObjectParamSetString(pCommand, "queueAction", "getStatus");
	if (argc == EXPECTED_MAX_NUM_ARGS)
	{
		aciObjectParamSetString(pCommand, "token", argv[4]);
	}
	aciObjectParamSetString(pCommand, "responseFormat", "xml");

	aciObjectToString(pCommand);

	printf("\nExecuting action...\n");

	aciError aci_status;
	
	// token is optional. if omitted then simply print entire response
	if (argc == EXPECTED_MAX_NUM_ARGS) 
	{		
		t_aciObject *pResult = NULL;
		aci_status = aciObjectExecute(pConnection, pCommand, &pResult);
		printf("Handling response...\n");


		if (aci_status == ACICONTENT_SUCCESS)
		{
			printf("Got response!\n");

			// Get <response> and print it
			t_aciObject* response = NULL;
			response = acioFindFirstOccurrence(pResult, "response");
			if (response)
			{
				char *szResponse = NULL;
				aci_status = acioParamGetString(response, ACI_DATA_NODE_VALUE, &szResponse);
				if (aci_status == ACICONTENT_SUCCESS) 
				{
					printf("response: %s\n", szResponse);
					
					// A successful communications response occurred.
					// Now need to need if the action response is ok.
					if (strcmp(szResponse, "SUCCESS")) 
					{
						// Get <errorstring> and print it
						t_aciObject* errorString = NULL;
						errorString = acioFindFirstOccurrence(pResult, "errorstring");
						if (errorString)
						{
							char *szErrorString = NULL;
							aci_status = acioParamGetString(errorString, ACI_DATA_NODE_VALUE, &szErrorString);
							if (aci_status == ACICONTENT_SUCCESS) 
							{
								printf("Error string: %s\n", szErrorString);
								exit(1);
							}
						}
					}
				} else {
					printf("Error! ACI Status: %d\n", aci_status);
					exit(1);
				}
			}
			
			// Get status and print it
			t_aciObject* status = NULL;
			status = acioFindFirstOccurrence(pResult, "status");
			if (status)
			{
				char *szStatus = NULL;
				aci_status = acioParamGetString(status, ACI_DATA_NODE_VALUE, &szStatus);
				if (aci_status == ACICONTENT_SUCCESS) 
				{
					printf("Queue Name: %s.\nToken: %s.\nStatus: %s\n", argv[3], argv[4], szStatus);
				} else {
					printf("Error! ACI Status: %d\n", aci_status);
					exit(1);
				}
			}
		} else {
			printf("Error! ACI Status: %d\n", aci_status);
			exit(1);
		}

		// Remember to tidy up afterwards
		aciObjectDestroy(&pResult);
	} else {
		char *szResponse = NULL;
		int nResponseLen = 0;
		char *szContentType = NULL;
		aci_status = aciObjectExecuteToString(pConnection, pCommand, &szResponse, &nResponseLen, &szContentType);

		if (aci_status == ACICONTENT_SUCCESS)
		{
			printf("Success!\n");
			
			// For most actions, szResponse now holds the raw xml and szContentType will be "application/xml"
			// Process the xml in whatever way you wish now (e.g. third-party XML processing API)
			printf("Response length: %d, ContentType: %s\n", nResponseLen, szContentType);
			printf("Response: %s\n", szResponse);
		} else {
			printf("Error! ACI Status: %d\n", aci_status);
			exit(1);
		}

		// Remember to tidy up afterwards
		free(szResponse);
		free(szContentType);
	}
	
	exit(0);
}
